src/completion.ml

```diff
@@ -0,0 +1,292 @@
+
+open Terms
+open Basics
+open Extra
+open Timed
+
+type 'a cmp = 'a -> 'a -> int
+
+(** [ord_lex ord] computes the lexicographic order corresponding to the
+    alphabetical order [ord] *)
+let rec ord_lex : 'a cmp -> 'a list cmp = fun ord l1 l2 ->
+  match (l1, l2) with
+  | [], []             -> 0
+  | [], _              -> -1
+  | _, []              -> 1
+  | h1 :: t1, h2 :: t2 ->
+      match ord h1 h2 with
+      | 0 -> ord_lex ord t1 t2
+      | x -> x
+
+let get_sym : term -> sym = fun t -> match unfold t with
+  | Symb (s, _) -> s
+  | _           -> assert false
+
+(** [lpo ord] computes the lexicographic path ordering corresponding to
+    the strict total order [ord] on the set of all function symbols *)
+let rec lpo : sym cmp -> term cmp = fun ord t1 t2 ->
+  let f, args = get_args t1 in
+  let f = get_sym f in
+  if List.exists (fun x -> lpo ord x t2 >= 0) args then 1
+  else
+    let g, args' = get_args t2 in
+    let g = get_sym g in
+    match ord f g with
+    | 1 ->
+        if List.for_all (fun x -> lpo ord t1 x > 0) args' then 1
+        else -1
+    | 0 ->
+        if List.for_all (fun x -> lpo ord t1 x > 0) args' then
+          ord_lex (lpo ord) args args'
+        else -1
+    | _ -> -1
+
+(** [build_meta_type k] builds the type “∀(x₁:A₁) (x₂:A₂) ⋯ (xk:Ak), B” where
+    “x₁” through “xk” are fresh variables, “Ai = Mi[x₁,⋯,x(i-1)]”, “Mi” is a
+    new metavar of arity “i-1” and type “∀(x₁:A₂) ⋯ (x(i-1):A(i-1), TYPE”. *)
+let build_meta_type : int -> term = fun k ->
+  assert (k>=0);
+  let vs = Bindlib.new_mvar mkfree (Array.make k "x") in
+  let rec build_prod k p =
+    if k = 0 then p
+    else
+      let k = k-1 in
+      let mk_typ = Bindlib.unbox (build_prod k _Type) in
+      let mk = fresh_meta mk_typ k in
+      let tk = _Meta mk (Array.map _Vari (Array.sub vs 0 k)) in
+      let b = Bindlib.bind_var vs.(k) p in
+      let p = Bindlib.box_apply2 (fun a b -> Prod(a,b)) tk b in
+      build_prod k p
+  in
+  let mk_typ = Bindlib.unbox (build_prod k _Type) (*FIXME?*) in
+  let mk = fresh_meta mk_typ k in
+  let tk = _Meta mk (Array.map _Vari vs) in
+  Bindlib.unbox (build_prod k tk)
+
+type rule = sym * Terms.rule
+
+(** [to_m k metas t] computes a new (boxed) term by replacing every pattern
+    variable in [t] by a fresh meta-variable and store the latter in [metas],
+    where [k] indicates the order of the term obtained *)
+let rec to_m : int -> (meta option) array -> term -> tbox = fun k metas t ->
+  match unfold t with
+  | Vari x         -> _Vari x
+  | Symb (s, h)    -> _Symb s h
+  | Abst (a, t)    ->
+      let (x, t) = Bindlib.unbind t in
+      _Abst (to_m 0 metas a) (Bindlib.bind_var x (to_m 0 metas t))
+  | Appl (t, u)    -> _Appl (to_m (k + 1) metas t) (to_m 0 metas u)
+  | Patt (i, n, a) ->
+      begin
+        let a = Array.map (to_m 0 metas) a in
+        let l = Array.length a in
+        match i with
+        | None   ->
+            let m = fresh_meta ~name:n (build_meta_type (l + k)) l in
```

Comment:
For confluence, we do not care of types. Simply use Kind instead of build_meta_type.

Reply:
In fact, this function is also used in sr.ml when we attempt to infer the type of the LHS of a rule. In this case, the types of the metavariables count (see term_of_meta in terms.ml).

Reply:
Right. But this was the point of a previous comment: don't copy and paste, but factorize! This function build_meta_type should be defined only once in Basics, and should be renamed by something more appropriate like build_prod_meta.